verilog/analysis/checkers/dff_name_style_rule.cc

```diff
@@ -0,0 +1,402 @@
+  absl::string_view base;
+  // If there are no patterns to check against, everything passes the check
+  if (suffixes.empty()) return base;
+
+  // Lhs and Rhs should be plain variable identifiers
+  const verible::SyntaxTreeLeaf *leftmost_leaf = verible::GetLeftmostLeaf(root);
+  const verible::SyntaxTreeLeaf *rightmost_leaf =
+      verible::GetRightmostLeaf(root);
+
+  // TODO: this is notably restrictive, we might want to allow things like
+  // `data_q[index] <= data_n`
+  if (leftmost_leaf != rightmost_leaf ||
+      leftmost_leaf->get().token_enum() != SymbolIdentifier) {
+    violations_.insert(LintViolation(
+        root,
+        absl::StrCat(
+            id, " Should be a simple reference, ending with a valid suffix: {",
+            absl::StrJoin(suffixes, ","), "}"),
+        context));
+
+    return base;
+  }
+
+  absl::string_view suffix_match;
+  // Check if id conforms to any valid suffix
+  const bool id_ok = std::any_of(suffixes.cbegin(), suffixes.cend(),
+                                 [&](const std::string &suffix) -> bool {
+                                   if (absl::EndsWith(id, suffix)) {
+                                     base = absl::string_view(
+                                         id.begin(), id.size() - suffix.size());
+                                     suffix_match = suffix;
+                                     return true;
+                                   }
+                                   return false;
+                                 });
+
+  // Exact match: id: "_q", suffix: "_q", there is no base
+  bool exact_match = suffix_match.size() == id.size();
+  if (id_ok && !exact_match) return base;
+
+  // The identifier can't be just the suffix we're matching against
+  const std::string lint_message =
+      exact_match
+          ? absl::StrCat(
+                "A valid identifier should not exactly match a valid prefix \"",
+                id, "\" == \"", suffix_match, "\"")
+          : absl::StrCat(id, " should end with a valid suffix: {",
+                         absl::StrJoin(suffixes, ","), "}");
+  violations_.insert(LintViolation(root, lint_message, context));
+
+  return base;
+}
+
+absl::Status DffNameStyleRule::Configure(absl::string_view configuration) {
+  // If configuration is empty, stick to the default
+  if (configuration.empty()) return absl::OkStatus();
+
+  // Default values
+  std::string output{kDefaultOutputSuffixes.data(),
+                     kDefaultOutputSuffixes.length()};
+  std::string input{kDefaultInputSuffixes.data(),
+                    kDefaultInputSuffixes.length()};
+  std::string waive_conditions_str{kDefaultWaiveConditions.data(),
+                                   kDefaultWaiveConditions.length()};
+
+  absl::Status status = verible::ParseNameValues(
+      configuration, {{"output", SetString(&output)},
+                      {"input", SetString(&input)},
+                      {"waive_regex", verible::config::SetRegex(&waive_regex)},
```

Comment:
Nice to see the regex configuration infrastructure @sconwayaus built in action here!

Reply:
Yeah, it is very nice :)